spi/qspi: add a testcase to test spi api for microchip board.

chen-png opened the pull request:
test write and read data with flash device using
spi single, dual, quad and octal mode.

for:  #27999

Signed-off-by: peng1 chen <peng1.chen@intel.com>

Commit:
spi/qspi: add a testcase to test spi api for microchip board.

test write and read data with flash device using
spi single, dual, quad and octal mode.

Signed-off-by: peng1 chen <peng1.chen@intel.com>

chen-png commented:
@ashaannajoy  this is a testcase to test qspi mode, i tested it on mec15xxevb_assy6853 and it was good. Could you take a look of it?

LeiW000 commented:
Compliance checking fails. I believe that we have to fix it first.

tbursztyka reviewed (changes requested):
So far, only xec driver seems to implement dual/quad lines support. And I am not sure it even works.

In fact, dual/quad/octal modes is not properly exposed in SPI API since it's far too basic and misleading as it only scratches the surface of proper dual/quad/octal mode support. Users won't be able to take advantage of these via the current API.

I made a proposal back then in #20564 which I believe would still be the way to go for generic SPI controllers exposing these modes. (there are flash optimized controllers using QSPI which do not need to be accessed via SPI API, these directly get a driver in flash area).

If you think that proposal would fit you need, I'll re-open it.

scottwcpg commented on tests/drivers/spi/qspi/src/main.c:
```diff
@@ -0,0 +1,375 @@
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	safbuf[4] = 0x0U;
```
Using transmit of 0 for dummy clocks can result in the controller and target both driving the pins at some point during the dummy clocks. This doesn't matter for single(full-duplex) mode because transmit and receive are on different pins. The driver provides for dummy clock generation using another struct spi_buf with buf=NULL and len=number of bytes that will generate the correct number of dummy clocks. 8 clocks * 2 bits/clock = 16 bits or 2 bytes.

chen-png replied:
I have changed to use spi_buf to set dummy clocks.

scottwcpg commented on tests/drivers/spi/qspi/src/main.c:
```diff
@@ -0,0 +1,375 @@
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	safbuf[4] = 0x0U;
```
Same comment as for dual.  Pins driven by controller and target during dummy clocks.

chen-png replied:
done

scottwcpg commented on tests/drivers/spi/qspi/src/main.c:
```diff
@@ -0,0 +1,375 @@
+		zassert_true(ret == 0,
+				"Write spi status2 QE=1 spi_transceive failure: "
+				"error %d", ret);
+	}
```
The test is not reading STATUS2 again to make sure QE was set by the write STATUS2 command.

chen-png replied:
added it to check again

chen-png commented:
> If you think that proposal would fit you need, I'll re-open it.

I read the long discussion whether to expand existing spi api to support quad mode, i will check it whether it's necessary from my side.

albertofloyd commented:
> So far, only xec driver seems to implement dual/quad lines support. And I am not sure it even works.
> 
> In fact, dual/quad/octal modes is not properly exposed in SPI API since it's far too basic and misleading as it only scratches the surface of proper dual/quad/octal mode support. Users won't be able to take advantage of these via the current API.
> 
> I made a proposal back then in #20564 which I believe would still be the way to go for generic SPI controllers exposing these modes. (there are flash optimized controllers using QSPI which do not need to be accessed via SPI API, these directly get a driver in flash area).
> 
> If you think that proposal would fit you need, I'll re-open it.

@tbursztyka are you saying test cases shouldn't be merged using existing API due API limitations? 
Shouldn't we move on with test as is and update if API changes?  Or you see a problem with test itself?

tbursztyka commented:
> Shouldn't we move on with test as is and update if API changes? Or you see a problem with test itself?

I don't see a problem with the test (though I did not try it, I don't have xec hw).
But the API is definitely way too limited for dual/quad/octal modes.

If you tell me the test works as is for xec, we'll merge it.

chen-png commented:
@albertofloyd  Could you help to run this test  on your microchip board, and check whether it meets your requirement? thanks.
and I tested this on my microchip board, it was good.

chen-png commented:
@ashaannajoy  could you take a look of it and help test it? I added the octal read test and quad write test.

albertofloyd commented:
> @albertofloyd Could you help to run this test on your microchip board, and check whether it meets your requirement? thanks.
> and I tested this on my microchip board, it was good.

@ashaannajoy have you tested this?

albertofloyd commented on tests/drivers/spi/qspi/src/main.c:
```diff
@@ -0,0 +1,672 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
Is this copyright intended?

chen-png replied:
i have updated it

albertofloyd commented on tests/drivers/spi/qspi/src/main.c:
```diff
@@ -0,0 +1,672 @@
+	txb[cnt].len = 4U;
+
+	/* set the dummy clocks */
+	if ((spi_opcode & 0xFF00U)) {
```
Double parenthesis not required

chen-png replied:
fixed it.

tbursztyka reviewed (changes requested):
I am still not convinced this is going to work for all dual/quad/octal use cases. 

My previous comment about DW and OC_SIMPLE drivers still stands: they do not support dual/quad/octal mode. XEC is the only driver that somehow implemented the support so far.

If you really need generic access to dual/quad/octal modes, please consider re-opening #20564

chen-png commented:
> I am still not convinced this is going to work for all dual/quad/octal use cases.

this is not going to work for all dual/quad/octal use cases, just for xec spi on microchip boards, because only them support this for now.

chen-png commented:
@dcpleung  could you take a look of this? thanks.

dcpleung commented:
Since this test requires a downstream SPI flash to be connected, it is not a generic SPI test. I suggest moving this to be under `tests/boards` and specifically saying this is testing the Microchip board.

chen-png commented:
> moving this to be under `tests/boards` and specifically saying this is testing the Microchip board.

@dcpleung  I moved this to tests/boards/ directory, and added a README to say this is for microchip board.

chen-png commented:
@tbursztyka  I moved this testcase to tests/board directory and added a readme to warn it just for microchip board, is it ok for you?
